src/lib.rs

```diff
@@ -1229,50 +1239,47 @@ struct ApiTable {
 
 #[link(name = "msquic")]
 extern "C" {
-    fn MsQuicOpenVersion(version: u32, api: &*const ApiTable) -> u32;
-    fn MsQuicClose(api: *const ApiTable);
+    fn MsQuicOpenVersion(version: u32, api: &*const ApiTableInner) -> u32;
+    fn MsQuicClose(api: *const ApiTableInner);
 }
 
 //
 // The following starts the "nice" Rust API wrapper on the C interop layer.
 //
 
 /// Top level entry point for the MsQuic API.
-///
-/// Developper must ensure a struct containing MsQuic members such as `Connection`
-///  or `Stream` declares `API` last so that the API is dropped last when the containing
-/// sruct goes out of scope.
+#[derive(Clone)]
 pub struct Api {
-    table: *const ApiTable,
+    table: Arc<ApiTable>,
 }
 
 /// The execution context for processing connections on the application's behalf.
 pub struct Registration {
-    table: *const ApiTable,
+    table: Arc<ApiTable>,
     handle: Handle,
 }
 
 /// Specifies how to configure a connection.
 pub struct Configuration {
-    table: *const ApiTable,
+    table: Arc<ApiTable>,
     handle: Handle,
 }
 
 /// A single QUIC connection.
 pub struct Connection {
-    table: *const ApiTable,
+    table: Arc<ApiTable>,
     handle: Handle,
 }
 
 /// A single server listener
 pub struct Listener {
-    table: *const ApiTable,
+    table: Arc<ApiTable>,
     handle: Handle,
 }
 
 /// A single QUIC stream on a parent connection.
 pub struct Stream {
-    table: *const ApiTable,
+    table: Arc<ApiTable>,
```

Comment:
Arc on every handle might have perf impact especially if sever uses many streams. Maybe we should follow the cpp wrapper to have a global API table: https://github.com/microsoft/msquic/blob/21e5b41e4e9f88d809de4c78e461b57641bfca8e/src/inc/msquic.hpp#L434

Reply:
The global API table may be better in the performance perspective. But it is difficult to free it in rust.

Reply:
Maybe it is a right use case for ctor: https://github.com/mmastrac/rust-ctor

Reply:
I've opend #4708.